File: backend/src/core/gamedata/types/material.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::serde_helpers::deserialize_fb_map;

// ============================================================================
// Enums
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub enum ItemRarity {
    #[serde(rename = "TIER_1")]
    #[default]
    Tier1,
    #[serde(rename = "TIER_2")]
    Tier2,
    #[serde(rename = "TIER_3")]
    Tier3,
    #[serde(rename = "TIER_4")]
    Tier4,
    #[serde(rename = "TIER_5")]
    Tier5,
    #[serde(rename = "TIER_6")]
    Tier6,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub enum ItemClass {
    #[serde(rename = "MATERIAL")]
    Material,
    #[serde(rename = "CONSUME")]
    Consumable,
    #[serde(rename = "NORMAL")]
    #[default]
    Normal,
    #[serde(rename = "NONE")]
    None,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
#[derive(Default)]
pub enum ItemType {
    Gold,
    CardExp,
    #[default]
    Material,
    Diamond,
    DiamondShd,
    HggShd,
    LggShd,
    ExpPlayer,
    PlayerAvatar,
    TktTry,
    TktryRecruit,
    TktInstFin,
    TktGacha,
    TktGacha10,
    SocialPt,
    ApGameplay,
    ApBase,
    TktGachaPrsv,
    LmtgsCoin,
    EpgsCoin,
    RepCoin,
    CrsShopCoin,
    CrsShopCoinV2,
    RetroCoin,
    RenamingCard,
    ApSupply,
    ExterminationAgent,
    LimitedTktGacha10,
    LinkageTktGacha10,
    VoucherPick,
    VoucherLevelmax6,
    VoucherLevelmax5,
    #[serde(rename = "VOUCHER_ELITE_II_6")]
    VoucherEliteIi6,
    #[serde(rename = "VOUCHER_ELITE_II_5")]
    VoucherEliteIi5,
    VoucherSkin,
    VoucherCgacha,
    OptionalVoucherPick,
    ItemPack,
    VoucherMgacha,
    VoucherFullPotential,
    UniCollection,
    ApItem,
    CrsRuneCoin,
    ActivityCoin,
    ActivityItem,
    EtStage,
    RlCoin,
    ReturnCredit,
    Medal,
    ActivityPotential,
    FavorAddItem,
    ClassicShd,
    ClassicTktGacha,
    ClassicTktGacha10,
    LimitedBuff,
    ClassicFesPickTier5,
    ClassicFesPickTier6,
    ReturnProgress,
    NewProgress,
    McardVoucher,
    MaterialIssueVoucher,
    SandboxToken,
    ExclusiveTktGacha,
    ExclusiveTktGacha10,
    #[serde(other)]
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "UPPERCASE")]
#[derive(Default)]
pub enum ItemOccPer {
    #[default]
    Usual,
    Almost,
    Always,
    Sometimes,
    Often,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "UPPERCASE")]
#[derive(Default)]
pub enum BuildingRoomType {
    #[default]
    Workshop,
    Manufacture,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
#[derive(Default)]
pub enum VoucherItemType {
    #[default]
    OptionalVoucherPick,
    MaterialIssueVoucher,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "UPPERCASE")]
#[derive(Default)]
pub enum VoucherDisplayType {
    #[default]
    None,
    Divide,
}

// ============================================================================
// Nested Structs
// ============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StageDrop {
    #[serde(alias = "StageId")]
    pub stage_id: String,
    #[serde(alias = "OccPer")]
    pub occ_per: ItemOccPer,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildingProduct {
    #[serde(alias = "RoomType")]
    pub room_type: BuildingRoomType,
    #[serde(alias = "FormulaId")]
    pub formula_id: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VoucherRelate {
    #[serde(alias = "VoucherId")]
    pub voucher_id: String,
    #[serde(alias = "VoucherItemType")]
    pub voucher_item_type: VoucherItemType,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UniqueItem {
    #[serde(alias = "Id")]
    pub id: String,
    #[serde(alias = "Count")]
    pub count: i32,
    #[serde(rename = "type", alias = "Type_")]
    pub item_type: String, // Almost always just "FURN"
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UniCollectionInfo {
    #[serde(alias = "UniCollectionItemId")]
    pub uni_collection_item_id: String,
    #[serde(alias = "UniqueItem")]
    pub unique_item: Vec<UniqueItem>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ItemPackContent {
    #[serde(alias = "Id")]
    pub id: String,
    #[serde(alias = "Count")]
    pub count: i32,
    #[serde(rename = "type", alias = "Type_")]
    pub item_type: ItemType,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ItemPackInfo {
    #[serde(alias = "PackId")]
    pub pack_id: String,
    #[serde(alias = "Content")]
    pub content: Vec<ItemPackContent>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FullPotentialCharacter {
    #[serde(alias = "ItemId")]
    pub item_id: String,
    #[serde(alias = "Ts")]
    pub ts: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityPotentialCharacter {
    #[serde(alias = "CharId")]
    pub char_id: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FavorCharacter {
    #[serde(alias = "ItemId")]
    pub item_id: String,
    #[serde(alias = "CharId")]
    pub char_id: String,
    #[serde(alias = "FavorAddAmt")]
    pub favor_add_amt: i32,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExpItem {
    #[serde(alias = "Id")]
    pub id: String,
    #[serde(alias = "GainExp")]
    pub gain_exp: i32,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApSupply {
    #[serde(alias = "Id")]
    pub id: String,
    #[serde(alias = "Ap")]
    pub ap: i32,
    #[serde(alias = "HasTs")]
    pub has_ts: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharVoucherItem {
    #[serde(alias = "Id")]
    pub id: String,
    #[serde(alias = "DisplayType")]
    pub display_type: VoucherDisplayType,
}

// ============================================================================
// Item
// ============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Item {
    #[serde(alias = "ItemId", default)]
    pub item_id: String,
    #[serde(alias = "Name", default)]
    pub name: String,
    #[serde(alias = "Description", default)]
    pub description: String,
    #[serde(alias = "Rarity", default)]
    pub rarity: ItemRarity,
    #[serde(alias = "IconId", default)]
    pub icon_id: String,
    #[serde(alias = "OverrideBkg")]
    pub override_bkg: Option<String>,
    #[serde(alias = "StackIconId")]
    pub stack_icon_id: Option<String>,
    #[serde(alias = "SortId", default)]
    pub sort_id: i32,
    #[serde(alias = "Usage", default)]
    pub usage: String,
    #[serde(alias = "ObtainApproach")]
    pub obtain_approach: Option<String>,
    #[serde(alias = "HideInItemGet", default)]
    pub hide_in_item_get: bool,
    #[serde(alias = "ClassifyType", default)]
    pub classify_type: ItemClass,
    #[serde(alias = "ItemType", default)]
    pub item_type: ItemType,
    #[serde(alias = "StageDropList", default)]
    pub stage_drop_list: Vec<StageDrop>,
    #[serde(alias = "BuildingProductList", default)]
    pub building_product_list: Vec<BuildingProduct>,
    #[serde(alias = "VoucherRelateList")]
    pub voucher_relate_list: Option<Vec<VoucherRelate>>,
}

// ============================================================================
// Container Types
// ============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Materials {
    pub items: HashMap<String, Item>,
    pub exp_items: HashMap<String, ExpItem>,
    pub potential_items: HashMap<String, HashMap<String, String>>,
    pub ap_supplies: HashMap<String, ApSupply>,
    pub char_voucher_items: HashMap<String, CharVoucherItem>,
}

// ============================================================================
// Table File Wrapper (for loading from FlatBuffer JSON)
// ============================================================================

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct ItemTableFile {
    #[serde(deserialize_with = "deserialize_fb_map")]
    pub items: HashMap<String, Item>,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub exp_items: HashMap<String, ExpItem>,
    #[serde(default)]
    pub potential_items: Vec<serde_json::Value>, // Complex nested structure with integer keys
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub ap_supply_out_of_date_dict: HashMap<String, ApSupply>,
    #[serde(deserialize_with = "deserialize_fb_map", default)]
    pub char_voucher_items: HashMap<String, CharVoucherItem>,
}

File: frontend/src/types/api/impl/material.ts
// Material types

export type ItemRarity = "TIER_1" | "TIER_2" | "TIER_3" | "TIER_4" | "TIER_5" | "TIER_6";

export type ItemClass = "MATERIAL" | "CONSUME" | "NORMAL" | "NONE";

export type ItemType =
    | "GOLD"
    | "CARD_EXP"
    | "MATERIAL"
    | "DIAMOND"
    | "DIAMOND_SHD"
    | "HGG_SHD"
    | "LGG_SHD"
    | "EXP_PLAYER"
    | "PLAYER_AVATAR"
    | "TKT_TRY"
    | "TKTRY_RECRUIT"
    | "TKT_INST_FIN"
    | "TKT_GACHA"
    | "TKT_GACHA_10"
    | "SOCIAL_PT"
    | "AP_GAMEPLAY"
    | "AP_BASE"
    | "TKT_GACHA_PRSV"
    | "LMTGS_COIN"
    | "EPGS_COIN"
    | "REP_COIN"
    | "CRS_SHOP_COIN"
    | "CRS_SHOP_COIN_V2"
    | "RETRO_COIN"
    | "RENAMING_CARD"
    | "AP_SUPPLY"
    | "EXTERMINATION_AGENT"
    | "LIMITED_TKT_GACHA_10"
    | "LINKAGE_TKT_GACHA_10"
    | "VOUCHER_PICK"
    | "VOUCHER_LEVELMAX_6"
    | "VOUCHER_LEVELMAX_5"
    | "VOUCHER_ELITE_II_6"
    | "VOUCHER_ELITE_II_5"
    | "VOUCHER_SKIN"
    | "VOUCHER_CGACHA"
    | "OPTIONAL_VOUCHER_PICK"
    | "ITEM_PACK"
    | "VOUCHER_MGACHA"
    | "VOUCHER_FULL_POTENTIAL"
    | "UNI_COLLECTION"
    | "AP_ITEM"
    | "CRS_RUNE_COIN"
    | "ACTIVITY_COIN"
    | "ACTIVITY_ITEM"
    | "ET_STAGE"
    | "RL_COIN"
    | "RETURN_CREDIT"
    | "MEDAL"
    | "ACTIVITY_POTENTIAL"
    | "FAVOR_ADD_ITEM"
    | "CLASSIC_SHD"
    | "CLASSIC_TKT_GACHA"
    | "CLASSIC_TKT_GACHA_10"
    | "LIMITED_BUFF"
    | "CLASSIC_FES_PICK_TIER_5"
    | "CLASSIC_FES_PICK_TIER_6"
    | "RETURN_PROGRESS"
    | "NEW_PROGRESS"
    | "MCARD_VOUCHER"
    | "MATERIAL_ISSUE_VOUCHER"
    | "SANDBOX_TOKEN"
    | "EXCLUSIVE_TKT_GACHA"
    | "EXCLUSIVE_TKT_GACHA_10"
    | string; // For unknown types

export type ItemOccPer = "USUAL" | "ALMOST" | "ALWAYS" | "SOMETIMES" | "OFTEN";

export type BuildingRoomType = "WORKSHOP" | "MANUFACTURE";

export type VoucherItemType = "OPTIONAL_VOUCHER_PICK" | "MATERIAL_ISSUE_VOUCHER";

export type VoucherDisplayType = "NONE" | "DIVIDE";

export interface StageDrop {
    stageId: string;
    occPer: ItemOccPer;
}

export interface BuildingProduct {
    roomType: BuildingRoomType;
    formulaId: string;
}

export interface VoucherRelate {
    voucherId: string;
    voucherItemType: VoucherItemType;
}

export interface UniqueItem {
    id: string;
    count: number;
    type: string;
}

export interface UniCollectionInfo {
    uniCollectionItemId: string;
    uniqueItem: UniqueItem[];
}

export interface ItemPackContent {
    id: string;
    count: number;
    type: ItemType;
}

export interface ItemPackInfo {
    packId: string;
    content: ItemPackContent[];
}

export interface FullPotentialCharacter {
    itemId: string;
    ts: number;
}

export interface ActivityPotentialCharacter {
    charId: string;
}

export interface FavorCharacter {
    itemId: string;
    charId: string;
    favorAddAmt: number;
}

export interface ExpItem {
    id: string;
    gainExp: number;
}

export interface ApSupply {
    id: string;
    ap: number;
    hasTs: boolean;
}

export interface CharVoucherItem {
    id: string;
    displayType: VoucherDisplayType;
}

export interface Item {
    itemId: string;
    name: string;
    description: string;
    rarity: ItemRarity;
    iconId: string;
    overrideBkg: string | null;
    stackIconId: string | null;
    sortId: number;
    usage: string;
    obtainApproach: string | null;
    hideInItemGet: boolean;
    classifyType: ItemClass;
    itemType: ItemType;
    stageDropList: StageDrop[];
    buildingProductList: BuildingProduct[];
    voucherRelateList: VoucherRelate[] | null;
}

export interface Materials {
    items: Record<string, Item>;
    expItems: Record<string, ExpItem>;
    potentialItems: Record<string, Record<string, string>>;
    apSupplies: Record<string, ApSupply>;
    charVoucherItems: Record<string, CharVoucherItem>;
}

export interface ItemTableFile {
    items: Record<string, Item>;
    expItems: Record<string, ExpItem>;
    potentialItems: unknown[];
    apSupplyOutOfDateDict: Record<string, ApSupply>;
    charVoucherItems: Record<string, CharVoucherItem>;
}
